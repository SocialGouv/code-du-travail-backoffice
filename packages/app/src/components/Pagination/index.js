import PropTypes from "prop-types";
import React from "react";

import { Container, NextButton, PageButton, PreviousButton } from "./index.style";

function renderPageButton({ onChange, pagesIndex, pagesLength }) {
  if (pagesLength <= 5) {
    return [...Array(pagesLength)].map((_, index) => (
      <PageButton
        color={index === pagesIndex ? "info" : "secondary"}
        data-testid={`button-${index + 1}`}
        key={index}
        onClick={() => (index !== pagesIndex ? onChange(index) : undefined)}
      >
        {index + 1}
      </PageButton>
    ));
  }

  const chunks = [];

  if (pagesIndex > 1) {
    chunks.push(
      <PageButton color="secondary" data-testid="button-1" key={0} onClick={() => onChange(0)}>
        1
      </PageButton>,
    );
  }

  if (pagesIndex > 2) {
    chunks.push(
      <span data-testid="separator-left" key="separator-left">
        …
      </span>,
    );
  }

  [...Array(3)].forEach((_, index) => {
    const pageIndex =
      pagesIndex === 0
        ? pagesIndex + index
        : pagesIndex === pagesLength - 1
        ? pagesIndex + index - 2
        : pagesIndex + index - 1;

    chunks.push(
      <PageButton
        color={pageIndex === pagesIndex ? "info" : "secondary"}
        data-testid={`button-${pageIndex + 1}`}
        key={pageIndex}
        onClick={() => (pageIndex !== pagesIndex ? onChange(pageIndex) : undefined)}
      >
        {pageIndex + 1}
      </PageButton>,
    );
  });

  if (pagesIndex < pagesLength - 3) {
    chunks.push(
      <span data-testid="separator-right" key="separator-right">
        …
      </span>,
    );
  }

  if (pagesIndex < pagesLength - 2) {
    chunks.push(
      <PageButton
        color="secondary"
        data-testid={`button-${pagesLength}`}
        key={pagesLength - 1}
        onClick={() => onChange(pagesLength - 1)}
      >
        {pagesLength}
      </PageButton>,
    );
  }

  return chunks;
}

function Pagination({ onChange, pagesIndex, pagesLength }) {
  return (
    <Container alignItems="flex-end" justifyContent="space-between">
      <PreviousButton
        color="secondary"
        data-testid="button-previous"
        icon="caret-left"
        isDisabled={pagesIndex === 0}
        onClick={() => (pagesIndex > 0 ? onChange(pagesIndex - 1) : undefined)}
      />
      {pagesLength !== 0 && <div>{renderPageButton({ onChange, pagesIndex, pagesLength })}</div>}
      <NextButton
        color="secondary"
        data-testid="button-next"
        icon="caret-right"
        isDisabled={pagesIndex === pagesLength - 1}
        onClick={() => (pagesIndex < pagesLength - 1 ? onChange(pagesIndex + 1) : undefined)}
      />
    </Container>
  );
}

Pagination.propTypes = {
  onChange: PropTypes.func.isRequired,
  pagesIndex: PropTypes.number.isRequired,
  pagesLength: PropTypes.number.isRequired,
};

export default Pagination;
