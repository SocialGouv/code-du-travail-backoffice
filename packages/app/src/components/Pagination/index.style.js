import styled from "@emotion/styled";
import { Flex } from "rebass";

import Button from "../../elements/Button";

export const Container = styled(Flex)`
  font-size: 1.25rem;
  margin-top: 1rem;
  min-height: 2rem;
  user-select: none;
`;

export const NextButton = styled(Button)`
  font-size: 1.25rem;

  > svg {
    margin: 0.325rem 0.9rem 0.325rem 1.1rem;
  }
`;
export const PreviousButton = styled(NextButton)`
  > svg {
    margin: 0.325rem 1.1rem 0.325rem 0.9rem;
  }
`;

export const PageButton = styled(Button)`
  font-size: 0.875rem;
  margin: 0 0.5rem;
`;
