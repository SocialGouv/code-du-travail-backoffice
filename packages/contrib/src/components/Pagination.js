import React from "react";
import ReactPaginate from "react-paginate";
import { Flex } from "rebass";
import styled from "styled-components";

const Container = styled(Flex)`
  user-select: none;
  margin: 1rem 0;

  ul {
    display: flex;
    list-style: none;
    margin-bottom: 0;
    padding-left: 0;
  }

  ul > li {
    a {
      color: inherit;
      cursor: pointer;
      font-size: 0.875rem;
      padding: 0.25rem 0.5rem;
      text-decoration: underline;
    }

    &.break {
      a {
        color: inherit;
        cursor: default;
      }
    }

    &.selected {
      a {
        color: inherit;
        cursor: default;
        font-weight: 600;
        text-decoration: none;
      }
    }
  }
`;

export default ({ initialPage, onPageChange, pageCount }) => (
  <Container justifyContent="center">
    <ReactPaginate
      disableInitialCallback={true}
      initialPage={initialPage}
      marginPagesDisplayed={3}
      nextLabel="Suivant >"
      onPageChange={onPageChange}
      pageCount={pageCount}
      pageRangeDisplayed={3}
      previousLabel="< Précédent"
    />
  </Container>
);
