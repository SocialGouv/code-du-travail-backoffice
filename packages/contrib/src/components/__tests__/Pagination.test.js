import React from "react";
import { fireEvent, render } from "react-testing-library";

import Pagination from "../Pagination";

describe("[Contrib] components/<Pagination />", () => {
  const props = {
    initialPage: 2,
    onPageChange: jest.fn(),
    pageCount: 12
  };

  const { container, getByText } = render(<Pagination {...props} />);

  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });

  it("should call onPageChange() with the expected param", async () => {
    fireEvent.click(getByText("2"));

    expect(props.onPageChange).toHaveBeenCalledWith({ selected: 1 });
  });
});
