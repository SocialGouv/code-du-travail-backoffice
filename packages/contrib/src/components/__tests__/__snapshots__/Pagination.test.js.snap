// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Contrib] components/<Pagination /> should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-bwzfXH Pagination__Container-sc-56xi-0 deNjEk"
  >
    <ul>
      <li
        class="previous"
      >
        <a
          aria-disabled="false"
          role="button"
          tabindex="0"
        >
          &lt; Précédent
        </a>
      </li>
      <li>
        <a
          aria-label="Page 1"
          role="button"
          tabindex="0"
        >
          1
        </a>
      </li>
      <li>
        <a
          aria-label="Page 2"
          role="button"
          tabindex="0"
        >
          2
        </a>
      </li>
      <li
        class="selected"
      >
        <a
          aria-current="page"
          aria-label="Page 3 is your current page"
          role="button"
          tabindex="0"
        >
          3
        </a>
      </li>
      <li>
        <a
          aria-label="Page 4"
          role="button"
          tabindex="0"
        >
          4
        </a>
      </li>
      <li
        class="break"
      >
        <a
          role="button"
          tabindex="0"
        >
          ...
        </a>
      </li>
      <li>
        <a
          aria-label="Page 10"
          role="button"
          tabindex="0"
        >
          10
        </a>
      </li>
      <li>
        <a
          aria-label="Page 11"
          role="button"
          tabindex="0"
        >
          11
        </a>
      </li>
      <li>
        <a
          aria-label="Page 12"
          role="button"
          tabindex="0"
        >
          12
        </a>
      </li>
      <li
        class="next"
      >
        <a
          aria-disabled="false"
          role="button"
          tabindex="0"
        >
          Suivant &gt;
        </a>
      </li>
    </ul>
  </div>
</div>
`;
